Revise SDL-0230: SPP resource management for Android (#770)
smartdevicelink/sdl_evolution: proposals/0230-spp-resource-management-for-android.md

### Detect the case where BluetoothServerSocket fails to accept a connection from head unit. 
Even though we could detect the case, we cannot increase the number of available SPP resources, because they are used by other apps.

> All we can do in this case is notify users that SPP channel runs out of available resources, and let users close some apps that may use the BluetoothSocket. It's not practical to show SPP service records that are used by Bluetooth adapter. The proposed approach is to notify users that we're running out of resources, and gives more information when user wants to.

Reviewer: ```suggestion
All we can do in this case is notify users that SPP channel runs out of available resources, and let users close some apps that may use the BluetoothSocket. It's not practical to show SPP service records that are used by Bluetooth adapter. The proposed approach is to notify users that we're running out of resources, and gives more information when the user requests it.
```

Prior to detecting the error, ```setState``` and ```notifyStateChanged``` in MultiplexBaseTransport class need to be extended, so that they can notify an error:
```java
    public static final String ERROR_REASON_KEY = "ERROR_REASON";
    public static final byte REASON_SPP_ERROR   = 0x01;    // REASON = SPP error, which is sent through Message.arg2.
    public static final byte REASON_NONE        = 0x0;

    protected synchronized void setState(int state, byte error) {
        if(state == mState){

>             return; //State hasn't changed. Will not updated listeners.

Reviewer: ```suggestion
            return; //State hasn't changed. Will not update listeners.
```

        }
        int arg2 = mState;
        mState = state;

        // Give the new state to the Handler so the UI Activity can update

>         //Also sending the previous state so we know if we lost a connection

Reviewer: ```suggestion
        //Also send the previous state so we know if we lost a connection
```

        notifyStateChanged(state, arg2, error);
    }

    private void notifyStateChanged(int arg1, int arg2, byte error) {
        Message msg = handler.obtainMessage(SdlRouterService.MESSAGE_STATE_CHANGE, arg1, arg2, getTransportRecord());
        Bundle bundle = new Bundle();
        bundle.putByte(ERROR_REASON_KEY, error);
        msg.setData(bundle);
        msg.sendToTarget();
    }
```

In MultiplexBluetoothTransport, stop method needs to be extended, so that the error information is passed to setState:
```java
    @Override
    protected synchronized void stop(int stateToTransitionTo, byte error) {
    	super.stop(stateToTransitionTo, error);
    	
    	...
        setState(stateToTransitionTo, error);
    }
```

The stop method gets called when server socket's accept failed:
```java

    try {
        ...
        socket = mServerSocket.accept();
    } catch(IOException e) {
        MultiplexBluetoothTransport.this.stop(STATE_ERROR, REASON_SPP_ERROR);
        return;
    }
```

> When TransportHandler in SdlRouterService detected STATE_ERROR, notifies it to user:

Reviewer: ```suggestion
When TransportHandler in SdlRouterService detects STATE_ERROR, notifies it to user:
```

The default language would be en_US.

Regarding who is responsible for the localization process, two steps are proposed:
1. English strings (which is default language) must be properly reviewed and maintained by steering committee.

> 2. Localized strings are maintained by project maintainer, and steering committee will be responsible for sign off those localized strings.  

Reviewer: ```suggestion
2. Localized strings are maintained by project maintainer, and steering committee will be responsible for sign off of those localized strings.  
```

### Listing application that uses Bluetooth

> When SPP resource error is detected, if we can narrow down the apps that actually uses the Bluetooth socket, that will be helpful for users to get recovered. Unfortunately, it is not realistic to identify exactly what apps are using Bluetooth. We can, however, list up applications that requests bluetooth permission, something like below:

Reviewer: ```suggestion
When SPP resource error is detected, if we can narrow down the apps that actually use the Bluetooth socket, that will be helpful for users to get recovered. Unfortunately, it is not realistic to identify exactly what apps are using Bluetooth. We can, however, list applications that request bluetooth permission, something like below:
```
